apollo-router/src/apollo_telemetry.rs

```diff
@@ -0,0 +1,409 @@
+                            client_version: span
+                                .attributes
+                                .get(&opentelemetry::Key::from_static_str("client_version"))
+                                .unwrap_or(&not_found)
+                                .to_string(),
+                        }),
+                        query_latency_stats: Some(QueryLatencyStats {
+                            latency_count: vec![busy_v],
+                            request_count: 1,
+                            ..Default::default()
+                        }),
+                        ..Default::default()
+                    };
+                    let operation_name = span
+                        .attributes
+                        .get(&opentelemetry::Key::from_static_str("operation_name"));
+                    // XXX NEED TO NORMALISE THE QUERY
+                    let key = normalize(operation_name, &q.as_str());
+
+                    let msg = reporter
+                        .submit_stats(graph.clone(), key, stats)
+                        .await
+                        .map_err::<TraceError, _>(|e| e.to_string().into())?
+                        .into_inner()
+                        .message;
+                    tracing::info!("server response: {}", msg);
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+fn normalize(op: Option<&opentelemetry::Value>, q: &str) -> String {
+    // If we don't have an operation name, no point normalizing
+    // it. Just return the unprocessed input.
+    let op_name: String = match op {
+        Some(v) => v.as_str().into_owned(),
+        None => return q.to_string(),
+    };
+    let parser = Parser::new(q);
+    // compress *before* parsing to modify whitespaces/comments
+    let ast = parser.compress().parse();
+    tracing::debug!("ast:\n {:?}", ast);
+    // If we can't parse the query, we definitely can't normalize it, so
+    // just return the un-processed input
+    if ast.errors().len() > 0 {
+        return q.to_string();
+    }
+    let doc = ast.document();
+    tracing::info!("{}", doc.format());
+    tracing::info!("looking for operation: {}", op_name);
+    let mut required_definitions: Vec<_> = doc
+        .definitions()
+        .into_iter()
+        .filter(|x| {
+            if let ast::Definition::OperationDefinition(op_def) = x {
+                match op_def.name() {
+                    Some(v) => return v.text() == op_name,
+                    None => return false,
+                }
+            }
+            false
+        })
+        .collect();
+    tracing::debug!("required definitions: {:?}", required_definitions);
+    assert_eq!(required_definitions.len(), 1);
+    let required_definition = required_definitions.pop().unwrap();
+    tracing::debug!("required_definition: {:?}", required_definition);
+    // XXX Somehow find fragments...
+    let def = required_definition.format();
+    format!("# {} \n{}", op_name, def)
```

Comment:
I want to note two subtle things about the operation name:

- It is based on the actual resolved operation name that is going to be executed.  GraphQL documents can contain one or more operation; if there is more than one they must all have names.  You may provide an operation name alongside the document; this is required if the document has more than one operation, and must match the operation if there's only one.  But you don't have to provide this out-of-band operation name if there's only one operation in the document even if the operation has a name. **In the case where a single named operation is provided and no name is provided alongside it, this key should still contain the operation name.**
- If the operation that is executing has no name (literally has no name like `query {` — not the same as "the GraphQL request doesn't specify a name alongside the document), then this operation name should be `-`, not the empty string.

Also there should be no space before the newline.

Reply:
My lack of GraphQL knowledge is hindering me here. I'll look further into this with the team to make sure I'm getting the right input and handling it correctly.

Reply:
Cool. BTW to be clear I didn't evaluate things in enough detail to know whether or not you were already doing the first bullet point properly.

And I spell this out in detail because I have a history of introducing bugs along these lines :) For example, Apollo Server has two fields `GraphQLRequest.operationName` and `GraphQLRequestContext.operationName`, and I thoroughly wish they were called `GraphQLRequest.explicitlyProvidedOperationName` and `GraphQLRequestContext.resolvedOperationName` instead of having the same name with subtly different semantics. I encourage you to differentiate these concepts by name when feasible.

Reply:
I believe, although I may be wrong, that the logic here is now doing the "right thing" when processing names. However, it's difficult to be sure until normalization is complete. I'll leave this unresolved for now.